### DISCUSSION TITLE
Breaking changes in minor releases

### DISCUSSION THREAD

**Discussion by dmitriy-kudelko** [twilio-labs/paste#3523] [Q&A] [answered]
Hey guys,

Our CI pipelines indicated some type errors in renovate bot PRs updating `@twilio-paste/core` from `20.3.0` to `20.4.0`.
It turned out that some types which were previously "public" and exposed from corresponding components are not exported any more. 

For example previously we relied on `SortDirection` exported from `@twilio-paste/core/data-grid`, but it's not there any more.
There are other similar cases in type incompatibilities as well.

Of course many of these issues could be fixed relatively easy by indirectly consuming corresponding wrapper types. e.g. `type SortDirection = DataGridHeaderSortProps['direction'];`. But it produces some work which needs to be done. And there's no mentioning about this in release notes AFAICS. Rather it gives an impression that the release is fully backward compatible.

So I'm wondering if introducing braking changes in minor releases is intended?

**Comment by SiTaggart** [comment, marked as answer]
Hi @dmitriy-kudelko 

Sorry you're experiencing some frustrations. Firstly, we don't intend to make breaking changes, but we also didn't intend to expose undocumented APIs from our packages. The public interface to each component is its Props Interface, and as you've already pointed out all the things you need exist on those publicly documented interfaces.

We knew it was going to be a risk as people do find unintentionally exposed private things, but I took the decision to mark it as a minor, as it's not technically anything we documented as a public interface. We are now moving to having the exported types stored in git so any small unintentional change to a resultant interface we'll know exactly what prop or HTML element is impacted.

> other similar cases in type incompatibilities as well.

I am interested to hear about these though, are there incompatibilities in the component interfaces that you've seen?

**Comment by dmitriy-kudelko** [comment]
Hi @SiTaggart , thanks for the explanation!

> We are now moving to having the exported types stored in git so any small unintentional change to a resultant interface we'll know exactly what prop or HTML element is impacted.

Sound good 👍 

In our case the migration turned out to be not that complex an we didn't rely on many undocumented props which were previously exported.

Things that we noticed:
- `useSideModalState` - `options` argument has become mandatory although previously it was optional
- the aforementioned `SortDirection`
- `TableAlignmentOptions` - we again had to indirectly access it via `TdProps['textAlign']`. Here things got a little bit tricky. `TdProps['textAlign']` is an optional prop and previously we did some mapping using `Record<TableAlignmentOptions, OurEnum>`, so we had to extract this type using `type TableAlignmentOptions = NonNullable<TdProps['textAlign']>;` to remove `undefined` from this type. 

Other that that the migration went smoothly.

**Reply by SiTaggart** [reply]
Excellent news! I'm so glad it wasn't so bad, and thanks for understanding
